spring-cloud-gcp-samples/spring-cloud-gcp-pubsub-config-sample/README.adoc

```diff
@@ -0,0 +1,78 @@
+The client uses the same binder with an anonymous subscription to receive notifications, and provides an HTTP endpoint `/message` for printing the dynamically updated message value.
+
+== Setup
+
+1. Configure your GCP project ID and credentials by following link:../../docs/src/main/asciidoc/core.adoc#project-id[these instructions].
++
+Alternatively, if you have the https://cloud.google.com/sdk/[Google Cloud SDK] installed and initialized, and are logged in with https://developers.google.com/identity/protocols/application-default-credentials[application default credentials], Spring will auto-discover those parameters for you.
+
+2. Make sure that GCP Cloud Pub/Sub API is enabled.
+
+== Running the code sample with configuration derived from local filesystem
+
+1. In separate terminal windows, start the `spring-cloud-gcp-pubsub-config-sample-server` and `spring-cloud-gcp-pubsub-config-sample-client` applications with the `$ mvn spring-boot:run` command in the same folder as the apps' `pom.xml` files.
+
+2. Visit http://localhost:8080/message in a browser; observe the default value of `none` printed on the page.
+This happens because there is no configuration available for config server yet.
+
+3. Create a temporary directory `/tmp/config`.
+Alternatively, update `spring.cloud.config.server.native.searchLocationsconfig` property in configuration server's link:spring-cloud-gcp-pubsub-config-sample-server/src/main/resources/application.properties[application.properties] with the directory you will use, and restart the config server.
+
+4. Place a file named `application.properties` with the following contents into the directory above:
++
+----
+example.message = message from the local filesystem
+----
+
+5. Notify the configuration server that new configuration is available.
+In a deployed server, this would be done through a GitHub webhook, or similar mechanism.
++
+----
+curl -X POST http://localhost:8888/monitor -d 'path=application'
+----
+
+6. Visit http://localhost:8080/message in a browser; observe the value now matches the value updated in the file system.
+
+== Running the code sample with configuration maintained in GitHub
+
+Keeping configuration in the local file system is simple, but not particularly useful -- if the configuration were available on the same file system as the client application, it could just get it without having to maintain a configuration server.
+A more realistic scenario is keeping configuration under version control.
+
+1. In the root of a GitHub repository, create a file named `application.properties` containing the following line:
++
+----
+example.message = message from GitHub
+----
+
+2. In configuration server's link:spring-cloud-gcp-pubsub-config-sample-server/src/main/resources/application.properties[application.properties]:
+  * Comment out `spring.profiles.active` and `spring.cloud.config.server.native.searchLocations` properties.
```

Comment:
Would it be easier to use two different config files and utilize spring profiles to switch between them?

Reply:
I was even thinking of having two separate modules, because I am planning to add the cloud-source-repositories sample as a separate module `spring-cloud-gcp-pubsub-config-server-csr`. So I could have modules like so:

````
spring-cloud-gcp-pubsub-config-server-local
spring-cloud-gcp-pubsub-config-server-github
spring-cloud-gcp-pubsub-config-server-csr
spring-cloud-gcp-pubsub-config-client
````

POM files for -csr would be different because it does not need `spring-cloud-config-monitor`. For the other two modules, only configuration would be different.

Reply:
I wouldn't mind, but if there is too much code duplication, it might not be worth it.

Reply:
For the server, it's just the spring boot app, plus `@EnableConfigServer` annotation. The license is way longer than the code. The mental load savings here are for the user not having to uncomment/recomment or synchronize the correct spring profile with the correct config file. I'll try multiple modules, if it looks verbose, we can always fall back to a single server.